src/shell/scripts/omp.lua

```diff
@@ -134,50 +182,60 @@ end
-        else
-            -- Create coroutine to generate tooltip rprompt.
-            cached_prompt.coroutine = coroutine.create(function ()
-                set_posh_tooltip(tip)
-                cached_prompt.tip_done = true
-                -- Refresh the prompt once the tooltip is generated.
-                clink.refilterprompt()
+
+    -- Get left/right prompts asynchronously, if possible.
+    if not cached_prompt.only_use_cache then
+        if can_async() then
+            -- IMPORTANT:  Defining this function inline makes sure it only
+            -- updates the same cached_prompt table that existed when the
+            -- function was defined.  That way if a new prompt starts (which
+            -- discards the old coroutine) and a new coroutine starts, the old
+            -- coroutine won't stomp on the new cached_prompt table.
+            clink.promptcoroutine(function ()
+                -- Generate left prompt, if needed.
+                if need_left then
+                    cached_prompt.left = get_posh_prompt(false)
+                end
+                -- Generate right prompt, if needed.
+                if rprompt_enabled then
+                    if need_left then
+                        -- Show left side while right side is being generated.
+                        display_cached_prompt()
+                    end
+                    cached_prompt.right = get_posh_prompt(true)
+                else
+                    cached_prompt.right = nil
+                end
             end)
+        else
+            if need_left then
+                cached_prompt.left = get_posh_prompt(false)
+            end
+            if rprompt_enabled then
+                cached_prompt.right = get_posh_prompt(true)
+            end
         end
-        if cached_prompt.tip_done then
-            -- Once the tooltip is ready, clear the Spacebar flag so that if the
-            -- tip changes and the Spacebar is pressed again, we can
-            -- generate a new tooltip.
-            cached_prompt.tip_done = nil
-            cached_prompt.tip_space = nil
-            cached_prompt.coroutine = nil
-        end
-    else
-        -- Tooltip is needed, but not in response to Spacebar, so refresh it
-        -- immediately.
-        set_posh_tooltip(tip)
     end
-    if not tooltip_active then
-        -- Tooltip is not active, generate rprompt normally.
-        cached_prompt.right = get_posh_prompt(true)
-    end
-    return cached_prompt.right, false
+
+    return cached_prompt.left
+end
+function p:rightfilter(prompt)
+    -- Return cached tooltip if available, otherwise return cached rprompt.
+    -- Returning false as the second return value halts further prompt
+    -- filtering, to keep other things from overriding what we generated.
+    return (cached_prompt.tooltip or cached_prompt.right), false
 end
 function p:transientfilter(prompt)
+-- REVIEW: this can potentially be expensive and noticable for users; consider
```

Comment:
what do you mean, cache the value once and reuse? This shouldn't be a very expensive prompt as the possibilities are limited to cached values already.

Reply:
The `transientfilter()` is called after the user presses <kbd>Enter</kbd> and it is not using a cached value.
It runs a new invocation of the OMP executable every single time.

Transient filters have the option to return whatever they want.  It's common to return something like simply `"> "`, but I don't know what range of possibilities OMP allows for there.

Reply:
Ah yes, the thing is, it can differ between runs. Instead of fetching live data, the transient prompt works with cached data from the previous run and resolves 1 template so it should be rather fast.

Reply:
Ok, so it should be relatively fast, but will be noticable when holding down Enter.

Do you mean the omp exe is caching between runs? That's what I was trying to suggest as the optimal approach for async prompt, and I thought you had said that was architecturally not feasible.

Reply:
@chrisant996 no it does not cache the prompt, it does cache all the results of every segment in the "normal", previous prompt run so it actually only renders a template.